config/src/main/java/org/springframework/security/config/annotation/method/configuration/PrePostAuthorizeBeanFactoryInitializationAotProcessor.java

```diff
@@ -0,0 +1,199 @@
+						.registerType(TypeReference.of(definition.getBeanClassName()),
+								MemberCategory.INVOKE_DECLARED_METHODS);
+				}
+				catch (NoSuchBeanDefinitionException ex) {
+					this.logger.debug(LogMessage.format(
+							"""
+									Could not register runtime hints for bean with name [%s] because it is not available, please provide
+									the needed hints manually""",
+							beanName));
+				}
+			}
+		}
+
+		private <A extends Annotation> List<A> collectAnnotations(Class<?> type, Class<A> annotationType) {
+			List<A> annotations = new ArrayList<>();
+			A classAnnotation = findDistinctAnnotation(type, annotationType, MergedAnnotation::synthesize);
+			if (classAnnotation != null) {
+				annotations.add(classAnnotation);
+			}
+			for (Method method : type.getDeclaredMethods()) {
+				A methodAnnotation = findDistinctAnnotation(method, annotationType, MergedAnnotation::synthesize);
+				if (methodAnnotation != null) {
+					annotations.add(methodAnnotation);
+				}
+			}
+			return annotations;
+		}
+
+		private Set<String> extractBeanNames(String rawExpression) {
+			SpelExpression expression = this.expressionParser.parseRaw(rawExpression);
+			SpelNode node = expression.getAST();
+			Set<String> beanNames = new HashSet<>();
+			resolveBeanNames(beanNames, node);
+			return beanNames;
+		}
+
+		private void resolveBeanNames(Set<String> beanNames, SpelNode node) {
+			if (node instanceof BeanReference br) {
+				beanNames.add(resolveBeanName(br));
+			}
+			int childCount = node.getChildCount();
+			if (childCount == 0) {
+				return;
+			}
+			for (int i = 0; i < childCount; i++) {
+				resolveBeanNames(beanNames, node.getChild(i));
+			}
+		}
+
+		private String resolveBeanName(BeanReference br) {
+			try {
+				Field field = ReflectionUtils.findField(BeanReference.class, "beanName");
+				field.setAccessible(true);
+				return (String) field.get(br);
+			}
+			catch (IllegalAccessException ex) {
+				throw new IllegalStateException("Could not resolve beanName for BeanReference [%s]".formatted(br), ex);
+			}
+		}
+
+		private static <A extends Annotation> A findDistinctAnnotation(AnnotatedElement annotatedElement,
+				Class<A> annotationType, Function<MergedAnnotation<A>, A> map) {
+			MergedAnnotations mergedAnnotations = MergedAnnotations.from(annotatedElement,
+					MergedAnnotations.SearchStrategy.TYPE_HIERARCHY, RepeatableContainers.none());
+			List<A> annotations = mergedAnnotations.stream(annotationType)
+				.map(MergedAnnotation::withNonMergedAttributes)
+				.map(map)
+				.distinct()
+				.toList();
+
+			return switch (annotations.size()) {
+				case 0 -> null;
+				case 1 -> annotations.get(0);
+				default -> throw new AnnotationConfigurationException("""
+						Please ensure there is one unique annotation of type @%s attributed to %s. \
+						Found %d competing annotations: %s""".formatted(annotationType.getName(), annotatedElement,
+						annotations.size(), annotations));
+			};
+		}
```

Comment:
@jzheaux This was copied from `AuthorizationAnnotationUtils`. My concern is that this logic can change and then we have to remember to update it on multiple places. Any idea on how we can reuse the same logic?